src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -589,166 +443,247 @@ void GrpcAresRequest::OnAresBackupPollAlarm() {
-    if (!(*it)->already_shutdown) {
+  for (const auto& fd_node : fd_node_list_) {
+    if (!fd_node->already_shutdown) {
       GRPC_ARES_WRAPPER_TRACE_LOG("request: %p shutdown fd: %s", this,
-                                  (*it)->polled_fd->GetName());
-      (*it)->polled_fd->ShutdownLocked(status);
-      (*it)->already_shutdown = true;
+                                  fd_node->polled_fd->GetName());
+      fd_node->polled_fd->ShutdownLocked(status);
+      fd_node->already_shutdown = true;
     }
   }
 }
 
-GrpcAresHostnameRequest::GrpcAresHostnameRequest(
+absl::StatusOr<GrpcAresHostnameRequest*> GrpcAresHostnameRequest::Create(
     absl::string_view name, absl::string_view default_port,
-    EventEngine::Duration timeout,
+    absl::string_view dns_server, EventEngine::Duration timeout,
     std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
-    EventEngine* event_engine)
-    : GrpcAresRequest(name, timeout, std::move(polled_fd_factory),
-                      event_engine),
-      default_port_(default_port) {}
-
-GrpcAresHostnameRequest::~GrpcAresHostnameRequest() {
-  // Destruction of on_resolve_ may trigger a chain of destruction that may
-  // require ExecCtx in the current thread.
-  grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
-  grpc_core::ExecCtx exec_ctx;
-  { auto c = std::move(on_resolve_); }
-  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p destructor", this);
-}
-
-void GrpcAresHostnameRequest::Start(
-    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
-  // Holds a ref across this function since OnResolve might be called inline
-  // inside ares_gethostbyname and we might be done in that case. This ref needs
-  // to be taken outside of the lock to prevent UAF during unlock if the
-  // destruction of the ref triggers the destruction of the request itself.
-  auto self = Ref(DEBUG_LOCATION, "Start");
-  grpc_core::MutexLock lock(&mu_);
-  GPR_ASSERT(initialized_);
-  on_resolve_ = std::move(on_resolve);
-  GRPC_ARES_WRAPPER_TRACE_LOG(
-      "request:%p c-ares GrpcAresHostnameRequest::Start name=%s, "
-      "default_port=%s",
-      this, name_.c_str(), default_port_.c_str());
-  // Early out if the target is an ipv4 or ipv6 literal.
-  if (ResolveAsIPLiteralLocked()) {
-    return;
-  }
-  // TODO(yijiem): Early out if the target is localhost and we're on Windows.
-
-  // We add up pending_queries_ here since ares_gethostbyname may directly
-  // invoke the callback inline e.g. if there is any error with the input. The
-  // callback will invoke OnResolve with an error status and may start the
-  // shutdown process too early (before the second ares_gethostbyname) if we
-  // haven't added up here.
-  ++pending_queries_;
+    EventEngine* event_engine) ABSL_NO_THREAD_SAFETY_ANALYSIS {
+  auto* request = new GrpcAresHostnameRequest(
+      name, default_port, timeout, std::move(polled_fd_factory), event_engine);
+  absl::StatusOr<std::string> result = request->ParseNameToResolve();
```

Comment:
Suggest calling this variable `port`.

Reply:
Done. (responding ahead of time since the next update will make this comment outdated)